### DISCUSSION TITLE
Add and Delete Sequences

### DISCUSSION THREAD

**Discussion by mvahowe** [Proskomma/epitelete#23] [Ideas]
Elsewhere, @mandolyte wrote:
> On handling deletion and addition of sequences...

> Suppose we followed the following rules:

> if the value for a sequence is the empty string, then remove it from the PERF
> if the value for a sequence Id does not exist in the PERF, then add it.

> Both these would need to be done prior to validation.

**Comment by mvahowe** [comment]
Thanks for this! The most immediate scenario is a bit different to those above ie
- sequence A includes a graft to sequence B
- that graft is deleted from sequence A
- sequence A is updated
- sequence B is now an orphan!

To spot this type of issue I think we need to implement something a garbage collection-like approach, ie
- collect all the graft ids, recursively, starting from main
- delete sequences that are not in that collection (and which are not the main sequence)

BUT Proskomma already does this. So maybe, once we get that bit of roundtripping working, we should let Proskomma tell Epitelete when this happens?

I think that removing the graft, as in the scenario above would be the natural way to remove a sequence.

I think that Proskomma also needs to be involved in creating new sequences. Otherwise, we need to check that, eg, the proposed sequenceId from the editor is not a duplicate (and, at that point, we might as well make our own). So I'm imagining some specific markup inside a graft, eg maybe a `new: true` flag and no `target`. At that point Epitelete
- asks Proskomma to make a new sequence of a given type and to return the id of that new sequence
- returns a new set of document data to the client, with that id dropped into the PERF graft
- editor can then potentially edit and write that new, empty sequence

Note that, in this scenario, the new sequence is empty but we wouldn't want it to be garbage collected!

Open to other ideas. I'm expecting this to need some schema changes, and I'm starting a discussion to collect those schema-related thoughts.

**Reply by mandolyte** [reply]
I'm a little unclear on how the editor takes all the sequences and orders them for viewing. Does each sequence reference its (non-graft) children? If it does, starting at the "root" main sequence, then it will never "see" the orphaned graft. So then we could leave it in there and let PK garbage collect it when "saved" (re-imported).
